File: day12/day12.kt
import java.io.File

class Ship(var tx: Int, var ty: Int, val movingTarget: Boolean) {
    var x: Int = 0
    var y: Int = 0
    
    fun move(cmd: Char, dist: Int) {
        when (cmd) {
            'N', 'S' -> {
                val dy = if (cmd == 'N') dist else -dist
                if (movingTarget) ty += dy else y += dy
            }
            'E', 'W' -> {
                val dx = if (cmd == 'E') dist else -dist
                if (movingTarget) tx += dx else x += dx
            }
            'L', 'R' -> {
                val px = tx; val py = ty
                val rad = (if (cmd == 'R') dist else -dist) * PI / 180
                tx = (px * cos(rad) + py * sin(rad)).roundToInt()
                ty = (-px * sin(rad) + py * cos(rad)).roundToInt()
            }
            'F' -> {
                x += tx * dist
                y += ty * dist
            }
        }
    }
    
    fun moveAll(input: List<String>): Pair<Int, Int> {
        for (line in input) move(line[0], line.drop(1).toInt())
        return Pair(x, y)
    }
}

fun main() {
    val input = File("day12.txt").readLines()
    val result = { t: Pair<Int,Int> -> t.first.absoluteValue + t.second.absoluteValue }
    println(result(Ship(1, 0, false).moveAll(input)))
    println(result(Ship(10, 1, true).moveAll(input)))
}

File: day12/day12.py
import math

class Ship:
    def __init__(self, dir_x, dir_y, target=False):
        self.x, self.y = 0, 0
        self.wx, self.wy = dir_x, dir_y
        self.target = target

    def move(self, line):
        cmd, dist = line[0], int(line[1:])
        if cmd in 'NS':
            dy = dist * (1 if cmd == 'N' else -1)
            if self.target: self.wy += dy
            else: self.y += dy
        elif cmd in 'EW':
            dx = dist * (1 if cmd == 'E' else -1)
            if self.target: self.wx += dx
            else: self.x += dx
        elif cmd in 'LR':
            angle = dist * (-1 if cmd == 'L' else 1)
            x, y, rad = self.wx, self.wy, angle * math.pi / 180
            self.wx = x * math.cos(rad) + y * math.sin(rad)
            self.wy = -x * math.sin(rad) + y * math.cos(rad)
        elif cmd == 'F':
            self.x += self.wx * dist
            self.y += self.wy * dist

    def move_all(self, data):
        for line in data: self.move(line)
        return self.x, self.y


data = list(open('day12.txt'))
result = lambda x: round(abs(x[0]) + abs(x[1]))
print(result(Ship(1, 0, False).move_all(data)))
print(result(Ship(10, 1, True).move_all(data)))
